Allow v1 search to use v2 auth with identity token

dmcgowan opened the pull request:
Updates the v1 search endpoint to also support v2 auth when an identity token is given. Only search v1 endpoint is supported since there is not v2 search currently defined to replace it.

Fixes #22379.

dmcgowan committed:
Vendor distribution changes

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
Tests related to identity tokens will be added to the distribution integration test package. Currently there is no registry setup and token server available for the Docker cli integration tests.

stevvooe commented on vendor/src/github.com/docker/distribution/registry/client/auth/session.go:
```diff
@@ -151,6 +155,19 @@ func (rs RepositoryScope) String() string {
 
+// RegistryScope represents a token scope for access
+// to resources in the registry.
+type RegistryScope struct {
```
Don't we have this definition in the distribution package?

dmcgowan replied:
This particular type was just added with https://github.com/docker/distribution/commit/145abeea7bb8a6bfe54d9cf05222caf9b8f7b80d.

Other areas in the code were just using `auth.Resource` for checking, https://github.com/docker/distribution/blob/641f1029677e49faa277f7769797518e973865fd/registry/handlers/app.go#L867. We want something with a stronger type for creating the scopes fro clients.

aaronlehmann commented on registry/service.go:
```diff
@@ -132,11 +133,44 @@ func (s *DefaultService) Search(ctx context.Context, term string, limit int, aut
+			v2Client.CheckRedirect = client.CheckRedirect
+			v2Client.Jar = client.Jar
+
+			logrus.Debugf("using v2 client")
```
Let's make this message more informative (make it clear this is for a search, maybe show the endpoint).

aaronlehmann commented:
Design LGTM

aaronlehmann commented on registry/auth.go:
```diff
@@ -168,6 +179,34 @@ func loginV2(authConfig *types.AuthConfig, endpoint APIEndpoint, userAgent strin
+
+	return &http.Client{
+		Transport: tr,
+		Timeout:   5 * time.Second,
```
Why not 15 seconds, since it gets overridden anyway?

dmcgowan replied:
Could use 15 seconds, I think 5 was used for other clients and it was increased for login. Given search may be slower anyway leaving at 15 makes sense.

aaronlehmann replied:
My point is that we seem to always overwrite this value (https://github.com/docker/docker/pull/24568/files#diff-2b07b8a4e93be324f246494ec743cde2R156 and https://github.com/docker/docker/pull/24568/files#diff-0ec52e4c7165f2740e01649dcef5867bR149), so it seems like we might as well use 15 seconds for simplicity to avoid one of those overwrites.

dmcgowan committed:
Allow v1 search to use v2 auth with identity token

Updates the v1 search endpoint to also support v2 auth when an identity token is given.
Only search v1 endpoint is supported since there is not v2 search currently defined to replace it.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

crosbymichael commented:
LGTM

thaJeztah commented:
ping @aaronlehmann PTAL

aaronlehmann commented:
Tested against hub while logged in and not logged in. I don't have a way of testing search with oauth. Based on the test in https://github.com/docker/distribution/issues/1838, LGTM.

thaJeztah commented:
LGTM, thanks

stevvooe commented on registry/auth.go:
```diff
@@ -215,7 +253,7 @@ func (err PingResponseError) Error() string {
 // whether v2 was confirmed by the response. If a response is received but
 // cannot be interpreted a PingResponseError will be returned.
-func PingV2Registry(endpoint APIEndpoint, transport http.RoundTripper) (auth.ChallengeManager, bool, error) {
+func PingV2Registry(endpoint *url.URL, transport http.RoundTripper) (auth.ChallengeManager, bool, error) {
```
Does this increase our reliance on the ping protocol, in any way? We really need to start understanding how to get rid of this, as it hampers the authentication model.

dmcgowan replied:
This does not increase our reliance on the ping protocol. Ping is used here the same way it is used for any v2 endpoint, to discover the token server. If we figure out a different way to do discovery or start caching this information, we will be able to take advantage of that here as well.
